transforms/React-DOM-to-react-dom-factories.js

```diff
@@ -0,0 +1,190 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+'use strict';
+
+module.exports = function(file, api, options) {
+  const j = api.jscodeshift;
+  const root = j(file.source);
+
+  let hasModifications;
+
+  const DOMModuleName = 'DOM';
+
+  const isDOMSpecifier = specifier => (
+    specifier.imported &&
+    specifier.imported.name === DOMModuleName
+  );
+
+  /**
+   * Replaces 'DOM' with 'createElement' in places where we grab 'DOM' out of
+   * 'React' with destructuring.
+   */
+  const replaceDestructuredDOMStatement = (j, root) => {
+    let hasModifications = false;
+    //---------
+    // First update import statments. eg:
+    // import {
+    //   DOM,
+    //   foo,
+    // } from 'react';
+    root
+      .find(j.ImportDeclaration)
+      .filter(path => (
+        path.node.specifiers.filter(isDOMSpecifier).length > 0 &&
+        path.node.source.value === 'react'
+      ))
+      .forEach(path => {
+        hasModifications = true;
+
+        // Replace the DOM key with 'createElement'
+        path.node.specifiers = path.node.specifiers.map(
+          specifier => {
+            if (specifier.imported && specifier.imported.name === DOMModuleName) {
+              return j.importSpecifier(j.identifier('createElement'));
+            } else {
+              return specifier;
+            }
+          }
+        );
+      });
+
+    //---------
+    // Next update require statments.
+    // This matches both
+    // const {
+    //   Component,
+    //   DOM,
+    // } = React;
+    // and
+    // const {
+    //   Component,
+    //   DOM,
+    // } = require('react');
+    root
+      .find(j.ObjectPattern)
+      .filter(path => (
+        path.parent.node.init &&
+        (
+          // matches '} = React;'
+          path.parent.node.init.name === 'React'
+          ||
+          (
+            // matches "} = require('react');"
+            path.parent.node.init.type === 'CallExpression' &&
+            path.parent.node.init.callee.name === 'require' &&
+            path.parent.node.init.arguments[0].value === 'react'
+          )
+        ) &&
+        path.node.properties.some(property => {
+          return property.key.name === DOMModuleName;
+        })
+      ))
+      .forEach(path => {
+        hasModifications = true;
+
+        // Replace the DOM key with 'createElement'
+        path.node.properties = path.node.properties.map((property) => {
+          if (property.key.name === DOMModuleName) {
+            return j.identifier('createElement');
+          } else {
+            return property;
+          }
+        });
+      });
+    return hasModifications;
+  };
+
+  hasModifications =
+    replaceDestructuredDOMStatement(j, root) || hasModifications;
+
+  const isDOMIdentifier = path => (
+    path.node.name === DOMModuleName &&
+    path.parent.parent.node.type === 'CallExpression'
```

Comment:
Glad you are thinking of that!

The way I structured the codemod, it only gets used after we have already detected the corresponding import/require statement. I can restructure things to make that more clear.

That's also why I added [this test.](https://github.com/flarnie/react-codemod/blob/26f661fc51504f4291cd70027b4f7d7b2b4d2d03/transforms/__testfixtures__/React-DOM-to-react-dom-factories/react-dom-no-change-dom-from-other-libraries.input.js)